daemon/api_systems.go

```diff
@@ -156,29 +157,42 @@ func getSystemDetails(c *Command, r *http.Request, user *auth.UserState) Respons
 	return SyncResponse(rsp)
 }
 
-type systemActionRequest struct {
-	Action string `json:"action"`
-	client.SystemAction
-}
-
 func postSystemsAction(c *Command, r *http.Request, user *auth.UserState) Response {
-	var req systemActionRequest
 	systemLabel := muxVars(r)["label"]
 
-	decoder := json.NewDecoder(r.Body)
-	if err := decoder.Decode(&req); err != nil {
-		return BadRequest("cannot decode request body into system action: %v", err)
+	// XXX: is there a better way to partially decode?
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return BadRequest("cannot read request body: %v", err)
+	}
+
+	var act struct {
+		Action string `json:"action"`
 	}
-	if decoder.More() {
-		return BadRequest("extra content found in request body")
+	if err := json.Unmarshal(body, &act); err != nil {
+		return BadRequest("cannot decode request body into system action: %v", err)
 	}
-	switch req.Action {
+	switch act.Action {
 	case "do":
+		var req client.SystemAction
+		if err := json.Unmarshal(body, &req); err != nil {
+			return BadRequest("cannot decode request body into system action: %v", err)
+		}
 		return postSystemActionDo(c, systemLabel, &req)
 	case "reboot":
+		var req client.SystemAction
+		if err := json.Unmarshal(body, &req); err != nil {
+			return BadRequest("cannot decode request body into system action: %v", err)
+		}
 		return postSystemActionReboot(c, systemLabel, &req)
+	case "install":
+		var req client.SystemActionInstall
```

Comment:
ah, it's used here but differently from SystemAction this not a result object

Reply:
I reworked this a bit but if you feel it should not be used here I can unexport and duplicate the fields.